### ISSUE TITLE
WeightNorm causes Memory Leak

### ISSUE THREAD

**Issue by MultiTrickFox** [pytorch/pytorch#73481]
### 🐛 Describe the bug

Long story short, calling modules which contain weight_norm causes the memory buildup over time - it is trivial to observe so I will not post the memory screenshots here. Tried on ubuntu latest & osx, my torch version is 1.10.2+cu113 (also happens in cpu version.)

it is very trivial code, but here is the exact piece of code that caused it. - bypass this module, no memory leak. do not bypass it, memory leak. It's simple really. And I'm 99% sure it's not from Conv1d..

weight_norm(
                    Conv1d(
                        channels,
                        channels,
                        kernel_size,
                        1,
                        dilation=dilation[0],
                        padding=get_padding(kernel_size, dilation[0]),
                    )
                ),

### Versions

PyTorch version: 1.10.2+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 21.10 (x86_64)
GCC version: (Ubuntu 11.2.0-7ubuntu2) 11.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.9.7 (default, Sep 10 2021, 14:59:43)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.13.0-30-generic-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 11.6.112
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1050
Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] pytorch-lightning==1.5.9
[pip3] torch==1.10.2+cu113
[pip3] torch-stft==0.1.4
[pip3] torchaudio==0.10.2+cu113
[pip3] torchmetrics==0.7.2
[pip3] torchtext==0.11.2
[pip3] torchvision==0.11.3+cu113
[conda] Could not collect

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

**Comment by albanD** [comment]
Hi,

Could you share the code you used to reproduce such behavior?

**Comment by MultiTrickFox** [comment]
The original code comes from YourTTS which is based on CoquiAI's TTS module. But I tracked it all the way from there to the weight_norm.py - bypassed all weight_norms, now no more memory buildup - just with that change.

My source is: 
https://colab.research.google.com/drive/1ftI0x16iqKgiQFgTjTDgRpOM1wC1U-yS

The function "synthesis" is where it occurs, and I went all the way down to weight_norm.

**Comment by albanD** [comment]
I am not familiar with this code so it is hard to figure out what is happening there.
The "synthesis" function seems to be doing the whole final creation of the wav so that's a lot of things in there.

Do you have a simple input for which calling weight_norm triggers this?

**Comment by MultiTrickFox** [comment]
Sorry I do not have a simpler code, but I'm pretty sure leak wouldn't be noticeable in a simple code, as this model is fairly big enough - and run consecutively, to make it noticeable.

But I could help in this way;
The modules which cause memory leak are located under:
TTS.vocoder.models.* and TTS.tts.models.*
The usage of "weight_norm" in these files is in the form of: 
self.something = nn.ModuleList([weight_norm(nn.Module(...)),])

I removed these exact weight_norms which proved them the cause for leak. The implementations of these files had no obvious mistakes to me, so I also opened the issue here. But if you're sure it's not on pytorch's end please feel free to close the issue here.. As an independent developer i already created a fix I can use right now. And thank you for your time..

**Comment by candlewill** [comment]
@MultiTrickFox Could you share the method of how did you fix the Memory Leak in WeightNorm? Btw, this problem is in the train or inference stage?

**Comment by MultiTrickFox** [comment]
I didn't test for training, just inference.
solution: torch.nn.utils.remove_weight_norm on all submodules.

I also found this method solves the problem only if you're sending the same input tensor over and over again !! only solves for this case.
I also found the conv2d's have a memory leak as well :) investigated further in pytorch forums, lots of more people talking on the same issue and pytorch officials directed to MKLDNN as the source of memory leaks. Long story short, I stopped when I saw that :) it's a deep rabbit hole..

Now I use a separate process for pytorch operations, then kill the process :) Not-so-good practice but it is the only way I come up with so far..
